[FLINK-12652] [documentation] add first version of a glossary

knaufk opened the pull request:
## What is the purpose of the change

* Part of FLIP-42
* Add a glossary section to the documentation for further reference and to establish a common terminology for common terms

## Brief change log

* add first version of glossary

## Verifying this change

* build docs and proof-read

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+
+#### Event
+
+An event is a statement of fact about a change of the state of the domain modelled by the
```
"Statement or fact"?

knaufk replied:
You're right this sounds too complicated. I changed it to just "statement".

NicoK replied:
ok, I first assumed a typo: `of`->`or`...but but it was really not. Then you're right :)

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+The term *instance* is used to describe a specific instance of a specific type (usually
+[Operator](./glossary#operator) or [Function](./glossary#function)) during runtime. As Apache Flink
+is mostly written in Java, this corresponds to the definition of *Instance* or *Object* in Java.
+In the context of Apache Flink, the term *parallel instance* is also frequently used to emphasizes
```
"to emphasize"

knaufk replied:
Done

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+
+#### Flink Job
+
+A Flink Job is the runtime representation of a Flink program. A Flink Job can be either submitted
```
"can either be..."

knaufk replied:
Done.

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+
+#### Managed State
+
+Managed State describes application state, which has been registered with the framework. For
```
No comma before `which`

knaufk replied:
Done

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+#### Managed State
+
+Managed State describes application state, which has been registered with the framework. For
+Managed State Apache Flink will take care about persistence and rescaling among other things.
```
"For Managed State,..." (here we need the comma)

knaufk replied:
Done

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+
+A physical graph is the result of translating a [Logical Graph](./glossary#logical-graph) for
+execution in a distributed runtime. The nodes are [Tasks](./glossary#task) and the edges indicate
+input/output-relationships or partitions of data streams or data sets.
```
link partition?

knaufk replied:
Done

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+
+#### Flink Session Cluster
+
+A long-running [Flink Cluster](./glossary#flink-cluster), which accepts multiple
```
No comma

knaufk replied:
Done

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+
+#### State Backend
+
+For stream processing programs the State Backend determines how state is stored on each Taskmanager
```
"For stream processing programs,..." (with comma)

knaufk replied:
Done

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+#### State Backend
+
+For stream processing programs the State Backend determines how state is stored on each Taskmanager
+(Java Heap of Taskmanager or RocksDB) as well as where it is written upon a checkpoint (Java Heap of Flink Master or
```
Maybe also emphasize "(embedded) RocksDB on TaskManager"?

knaufk replied:
Done

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+
+#### Sub-Task
+
+A Sub-Task is a [Task](./glossary#task) responsible for processing a
```
"is an instance of a Task"?

knaufk replied:
Done

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+
+A Sub-Task is a [Task](./glossary#task) responsible for processing a
+[partition](./glossary#partition) of the data stream. The term "Sub-Task" emphasizes that there are
+multiple parallel Tasks for the same [Operator](./glossary#operator).
```
"operator(s)"

knaufk replied:
Done

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+output new data streams or data sets. A transformation might change a data stream or data set on a
+per-record basis, but but might also only change its partitioning or perform an aggregation. In
+contrast to operations, transformations operate on streams or data sets, while operations operate
+on [records](./glossary#Record).
```
I don't really get the last argument. Also, do you compare it with Operators, or Operations?

knaufk replied:
For me, an operation would be the action that an operator performs. 

The idea is that transformations are applied on streams, while operations are applied on individual records.

NicoK replied:
Can you give an example of a "transformation" and an "operation"? Maybe I'm mixing things up here...

knaufk replied:
A transformation and and operation can do exactly the same thing. For example, a simple Map can be an operation if you talk about it on a per-record basis and a transformation if you talk about it on a stream.

For a repartitioning operations it makes a bit more sense, I think. A `keyBy` results is a transformation of the stream, but you can not repartition a single record. Of course, you could view it as a per-record operation, but it makes less sense, I think.

NicoK replied:
That makes things clearer, but do we actually need the two different terms? I think, it may actually confuse people (like me) - unless more specifically explained as here.

knaufk replied:
I changed this now. Since we don't mention the term "operation" anywhere, I removed this sentence and just noted that a transformation usually, but not always corresponds to an operator.

knaufk commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+#### Flink Cluster
+
+The distributed system consisting of (typically) one Flink Master process and one or more Flink
+Taskmanagers processes.
```
In my opinion, we should use "Flink Master (process)" instead of Flink JobManager, yes. Flink JobManager is too overloaded, since the FLIP-6 effort. Are there cases, where the JobManagers are not running in the same process as Resource Manager and Dispatcher?

I also thought about adding these to the Glossary. We should also add JobManager then, to avoid confusion, so: Flink Master, JobManager, TaskManager. 

What do you think?

NicoK replied:
In theory, yes: JobManager, ResourceManager and Dispatcher can be deployed separately...Flink's own startup scripts are not (yet?) making use of this though, afaik.

Unless you want to introduce new terminology, I think, we should stick with JobManager (process) even though it may contain more components.

knaufk replied:
I think, I would like to introduce new terminology ;) If I understood @StephanEwen correctly, he was also in favor of new terminology, but it is of course tricky.

knaufk replied:
@NicoK I changed this for "JobManager Process" for now. Let's move the discussion about renaming this to the appropriate PR for the Concepts or Deployment Section.

uce replied:
FWIW, I like `Flink Master Process` for the reasons outlined by Konstantin. I think it's a good call to move the discussion of whether we want to change it to another place/time.

fhueske replied:
just FYI, I also used "Flink master process" when describing the concept in the book.

knaufk commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,140 @@
+A Flink Application Cluster is a dedicated [Flink Cluster](./glossary#flink-cluster) that only
+executes a single [Flink Job](./glossary#flink-job). The lifetime of the
+[Flink Cluster](./glossary#flink-cluster) is bound to the lifetime of the Flink Job. Formerly
+Flink Application Clusters were also known as Flink Clusters in *job mode*.
```
Done

knaufk commented:
@NicoK  Thank you for your review. I have addressed all your comments and added `JobManager`, `Flink Master` and `TaskManager` to the glossary. 

I am not sure about operator chaining. Is it used at a lot of different places or is there confusion about what it means? (These were my main criteria, but these are also arbitrary, of course.)

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,158 @@
+
+The Flink Master is the master process of a [Flink Cluster](./glossary#flink-cluster). It consists
+of three main components: resource manager, dispatcher and one [JobManager](./glossary#job-manager)
+per running [Flink Job](./glossary#flink-job).
```
irrespective of whether to use "JobManager" or "Flink Master", reading about 3 components and then having "on JM per job" sounds a bit strange

knaufk replied:
three "different" components insteads?

NicoK replied:
yes, that might work

knaufk replied:
Done.

rmetzger commented:
@flinkbot approve description consensus

uce commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,166 @@
+under the License.
+-->
+
+#### Flink Application Cluster
```
Personally, I call these `job cluster` in line with the internals, e.g. [JobClusterEntrypoint](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/JobClusterEntrypoint.java).

cc @tillrohrmann

tillrohrmann replied:
It is the same for me. I'm also fine with changing the terminology as long as we don't confuse our users.

knaufk replied:
The term "Flink Application (Cluster)" was proposed by @StephanEwen to emphasize, that you basically run a Flink job like an classical application in this mode (or at least this is the mid-term goal (passive resource manager, Flink as a library). 

@tillrohrmann @uce Do you think we can change this at this point?

uce replied:
I do like the proposed name and I'd be happy to call it application cluster going forward myself.

I think having slight misalignment with internal class naming is acceptable, but I'd suggest to update user facing references to job cluster if we keep the renaming:
* https://ci.apache.org/projects/flink/flink-docs-release-1.8/ops/deployment/kubernetes.html#flink-job-cluster-on-kubernetes
* https://github.com/apache/flink/tree/master/flink-container/docker
* https://github.com/apache/flink/tree/master/flink-container/kubernetes

Otherwise I would be afraid of ending up in a situation with inconsistent names which can be confusing to users.

Since I am not involved in the day to day development of this component, I would leave the decision to @tillrohrmann whether we can change it at this point.

knaufk replied:
Agreed. There are actually a few changes "planned" to this module anyway (see discussion of https://issues.apache.org/jira/browse/FLINK-12546), so this might actually be good timing to rename it.

knaufk replied:
@tillrohrmann is fine to change this. @uce I created https://issues.apache.org/jira/browse/FLINK-12885 to rename Job Cluster to Application Cluster. Shouldn't be a problem to do this before the 1.9 release.

knaufk replied:
I have already a local version of FLINK-12885 now, but it is still blocked on a different issue (FLINK-12752). Once this is merged, we can align this pretty quickly.

uce commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,166 @@
+
+For stream processing programs, the State Backend determines how state is stored on each Taskmanager
+(Java Heap of Taskmanager or (embedded) RocksDB) as well as where it is written upon a checkpoint
+(Java Heap of Flink Master or Filesystem).
```
```suggestion
(Java Heap of Flink JobManager or Filesystem).
```

knaufk replied:
Done

fhueske commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,166 @@
+
+An event is a statement about a change of the state of the domain modelled by the
+application. Events can be input and/or output of a stream or batch processing application.
+Events are special types of [records](./glossary#Record)
```
Add `.` at the end

fhueske commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,166 @@
+
+Node of a [Physical Graph](./glossary#physical-graph). A task is the basic unit of work, which is
+executed by Flink's runtime. Tasks encapsulate exactly one parallel
+[Operator Chain](./glossary#operator-chain).
```
"one parallel Operator Chain" -> "one parallel instance of an Operator or Operator Chain"?

knaufk replied:
Done.

fhueske commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,166 @@
+
+#### Operator Chain
+
+An Operator Chain consists of one or more consecutive [Operators](./glossary#operator) without any
```
I would not call a single operator an operator chain. IMO a chain consist of at least two operators.

knaufk replied:
Ok, personally I think a chain of size 1 is perfectly fine, but I'll change it ;)

fhueske commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,166 @@
+#### Flink TaskManager
+
+TaskManagers are the worker processes of a [Flink Cluster](./glossary#flink-cluster).
+[Tasks](./glossary#task) are scheduled to TaskManagers for execution.
```
Mention that TMs communicate with each other to exchange data between tasks?

knaufk replied:
Done.

fhueske commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,166 @@
+TaskManagers are the worker processes of a [Flink Cluster](./glossary#flink-cluster).
+[Tasks](./glossary#task) are scheduled to TaskManagers for execution.
+
+#### Transformation
```
Mention that a Transformation is an API concept?

knaufk replied:
Can you elaborate?

fhueske replied:
From the description I read that the term "Transformation" is used in the context of the API, i.e., when implementing an application. When the application is executed it is rather operator or task. If this distinction is not intended, how would a transformation relate to operator/task/function?

knaufk replied:
Okay, I see. I have added a sentence. Let me know, if this fits.

knaufk commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,166 @@
+#### Flink JobManager Process
+
+The Job Manager Process is the master of a [Flink Cluster](./glossary#flink-cluster). It is called
+*JobManager* for historical reasons, but actually has actually contains three distinct components:
```
Currently, they are always run together, correct? I assume the other option would be: 

* Flink Master Process: Dispatcher &  Resource Manager
* One JobManager Process per JobManager separately

How about we it like it is right now and revisit this, when we have decided about introducing the term "Flink Master Process" in the course of the FLIP?

uce replied:
SGTM.

fhueske replied:
AFAIK, they are not always run together. At least that's what is described by FLIP-6. But I don't want to nitpick on this

knaufk replied:
@fhueske In which case are these not run together? AFAIK also in YARN/Mesos session mode, there is only one Master Process for all Job Managers, although conceptually they would not have to run together (since FLIP-6).

knaufk replied:
I talked to @tillrohrmann offline and they are currently always run together in the same process + there are no plans to change this in the foreseeable future.

knaufk commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,166 @@
+
+#### Function
+
+Functions, or user-defined functions (UDFs), are implemented by the user and encapsulate the
```
Good point. Done.

knaufk commented:
@uce @tillrohrmann @fhueske Thanks a lot for your comments. I addressed & resolved most of the points raised. Could you please have a look at the open discussion.

uce commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,167 @@
+
+#### Flink JobManager Process
+
+The Job Manager Process is the master of a [Flink Cluster](./glossary#flink-cluster). It is called
```
```suggestion
The JobManager Process is the master of a [Flink Cluster](./glossary#flink-cluster). It is called
```

knaufk replied:
Done.

uce commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,167 @@
+#### Flink JobManager Process
+
+The Job Manager Process is the master of a [Flink Cluster](./glossary#flink-cluster). It is called
+*JobManager* for historical reasons, but actually has actually contains three distinct components:
```
```suggestion
*JobManager* for historical reasons, but actually contains three distinct components:
```

knaufk replied:
Done.

fhueske commented on log:
```diff
@@ -0,0 +1,787 @@
+  1676	Till Rohrmann
```
This file and `.attach_pid22989` seem to be accidentally included in the PR.

knaufk replied:
Hoppala. Yes.

knaufk commented:
@uce @fhueske I talked to Till and addressed all open discussions. Any open points from your side?

knaufk commented:
@NicoK @fhueske Can we merge this?

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+#### Flink Cluster
+
+The distributed system consisting of (typically) one Flink Master process and one or more Flink
+Taskmanagers processes.
```
`Taskmanagers processes` -> `TaskManager processes`?

actually, the capitalisation should be unified throughout the glossary, sometimes you use `Taskmanager` instead)

knaufk replied:
Done.

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+
+An event is a statement about a change of the state of the domain modelled by the
+application. Events can be input and/or output of a stream or batch processing application.
+Events are special types of [records](./glossary#Record).
```
I find this rather complex: `a change of the state of the domain modelled by the
application` and honestly don't know what that means. Maybe rephrase but also give an example (not just "special type or record")

knaufk replied:
In Domain Driven Design, Event-Driven Architectures, etc. this is exactly, what an event is. It is always an event (=state change) in the domain. I could add a text book example like "checkout" or "item added to cart". Would this help?

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+#### Flink JobManager
+
+JobManagers are one of the components running in the
+[Flink JobManger Process](./glossary#flink-jobmanager-process). A JobManager is responsible for
```
`JobManger` -> `JobManager`

knaufk replied:
Done.

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+
+#### State Backend
+
+For stream processing programs, the State Backend determines how state is stored on each Taskmanager
```
`TaskManager`

knaufk replied:
Done.

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+#### State Backend
+
+For stream processing programs, the State Backend determines how state is stored on each Taskmanager
+(Java Heap of Taskmanager or (embedded) RocksDB) as well as where it is written upon a checkpoint
```
`TaskManager`

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+
+A Sub-Task is an instance of [Task](./glossary#task) responsible for processing a
+[partition](./glossary#partition) of the data stream. The term "Sub-Task" emphasizes that there are
+multiple parallel Tasks for the same [Operator](./glossary#operator).
```
`[Operator](./glossary#operator) or [Operator Chain](./glossary#operator-chain)`

knaufk replied:
Done.

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+
+A physical graph is the result of translating a [Logical Graph](./glossary#logical-graph) for
+execution in a distributed runtime. The nodes are [Tasks](./glossary#task) and the edges indicate
+input/output-relationships or [partitions](./glossary#partition) of data streams or data sets.
```
"physical graph" somehow implies this is the runtime graph of task instances on the nodes, but this is still only a data structure of how to deploy things, right?

knaufk replied:
That's a good question and also related to your other comment below. I would have said that the "physical graph" indeed is the "runtime graph" of task instances. Of course it is just a model of the actual physical graph, but it models the execution during runtime.

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+
+#### Sub-Task
+
+A Sub-Task is an instance of [Task](./glossary#task) responsible for processing a
```
`of a [Task]...`

knaufk replied:
Done.

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+
+Node of a [Physical Graph](./glossary#physical-graph). A task is the basic unit of work, which is
+executed by Flink's runtime. Tasks encapsulate exactly one parallel instance of an
+[Operator](./glossary#operator) or [Operator Chain](./glossary#operator-chain).
```
You got me confused a but with the last sentence thinking that a Task is also an instance. How about this:
`Tasks logically encapsulate exactly one parallel instance...`

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+
+A Transformation is applied on one or more data streams or data sets and results in one or more
+output data streams or data sets. A transformation might change a data stream or data set on a
+per-record basis, but but might also only change its partitioning or perform an aggregation. While
```
`but but` -> `but`

knaufk replied:
Done.

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+A Transformation is applied on one or more data streams or data sets and results in one or more
+output data streams or data sets. A transformation might change a data stream or data set on a
+per-record basis, but but might also only change its partitioning or perform an aggregation. While
+[Operators](./glossary#operator) and [Functions](./glossary#function)) are "physical" parts Flink's
```
`are the "physical" parts of Flink's`

knaufk replied:
Done.

NicoK commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,168 @@
+output data streams or data sets. A transformation might change a data stream or data set on a
+per-record basis, but but might also only change its partitioning or perform an aggregation. While
+[Operators](./glossary#operator) and [Functions](./glossary#function)) are "physical" parts Flink's
+API, Transformations are only an API concept. Specifically, most - but not all -  transformations
```
there is one space too many after the last "-"

knaufk replied:
Done.

alpinegizmo commented:
For a first version, this looks quite good already. Are we collecting somewhere a list of terms that might be added at some point? Job scheduling and back pressure come to mind as examples.

knaufk commented:
@NicoK Thanks a lot for the review. I have addressed your comments.

knaufk commented:
> For a first version, this looks quite good already. Are we collecting somewhere a list of terms that might be added at some point? Job scheduling and back pressure come to mind as examples.

@alpinegizmo Good point. I also have a view terms on my backlog (e.g. "snapshot", "savepoint" & "checkpoint"). I have created a Jira ticket https://issues.apache.org/jira/browse/FLINK-13002 to collect more terms.

1u0 commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,169 @@
+
+#### Flink Cluster
+
+The distributed system consisting of (typically) one Flink Master process and one or more Flink
```
Suggestion: 
 * `The distributed system ...` -> `A distributed system ...`
 * `Flink Master process` -> `[Flink JobManager process](#flink-jobmanager-process)`
 * `Flink TaskManager processes` -> `[Flink TaskManager processes](#...)`

knaufk replied:
2) and 3) done. 1) I am not sure.Maybe let's ask a native speaker. @sjwiesman what sounds more natural to you?

sjwiesman replied:
"A distributed system"

sjwiesman replied:
Also, drop the (typically) ?

knaufk replied:
Standby JobManager Processes? I would considers these to be part of the Flink Cluster.

sjwiesman replied:
Gotcha, maybe as a nit "The distributed system consisting of one Flink Master process, zero or more stand by job manager processes, and one or more Flink . . ."

Otherwise, lets get this merged in. We can always open follow up pull requests.

1u0 commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,169 @@
+#### Operator Chain
+
+An Operator Chain consists of two or more consecutive [Operators](./glossary#operator) without any
+repartitioning in between. Operators within the same Operation Chain forward records to each other
```
Typo: `Operation Chain` -> `Operator Chain`.

knaufk replied:
Done

1u0 commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,169 @@
+
+#### State Backend
+
+For stream processing programs, the State Backend determines how state is stored on each TaskManager
```
`the State Backend determines how state is stored` -> `a State Backend determines how Flink Job's state is stored`?

In general, it's a little bit not clear from here, that a state backend can be configured on Flink job level.

knaufk replied:
Should be clearer now.

1u0 commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,169 @@
+executed by Flink's runtime. Tasks logically encapsulate exactly one parallel instance of an
+[Operator](./glossary#operator) or [Operator Chain](./glossary#operator-chain).
+
+#### Flink TaskManager
```
Can be `Flink TaskManager Process` to be similar with `Flink JobManager Process`.

knaufk replied:
We only introduced the term "Flink JobManager Process" to differentiate between "Flink JobManager" (a component) of the "Flink JobManager Process". I think for TaskManager we don't need this.

knaufk replied:
As discussed with Till and Stephan we went for "Flink Master" after all.

1u0 commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,169 @@
+output data streams or data sets. A transformation might change a data stream or data set on a
+per-record basis, but might also only change its partitioning or perform an aggregation. While
+[Operators](./glossary#operator) and [Functions](./glossary#function)) are the "physical" parts
+Flink's API, Transformations are only an API concept. Specifically, most - but not all -
```
`Flink's API` -> `of Flink's API`?

knaufk replied:
Done

1u0 commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,169 @@
+per-record basis, but might also only change its partitioning or perform an aggregation. While
+[Operators](./glossary#operator) and [Functions](./glossary#function)) are the "physical" parts
+Flink's API, Transformations are only an API concept. Specifically, most - but not all -
+transformations are implemented by a certain [Operator](./glossary#operator).
```
`by a certain [Operator]` -> `by certain [Operators]`?

knaufk replied:
Done

knaufk commented:
> What about removing `./glossary` prefix in the references? I think they still should work and the document would be more rename-proof.

Indeed. Thanks! Done that.

knaufk commented:
@1u0 Thanks for taking a look! I have addressed your comments.

sjwiesman commented:
This looks really great @knaufk . If no one sees any glaring issues lets go ahead and get this merged in. Part of the reason for adding the glossary early is that we know we won't get everything right away and so once the page is merged in we can continue to incrementally improve it.

knaufk commented:
I have added a placeholder for the chinese version and created https://issues.apache.org/jira/browse/FLINK-13037 to translate it.

StephanEwen commented on docs/concepts/glossary.md:
```diff
@@ -0,0 +1,167 @@
+#### Flink Cluster
+
+A distributed system consisting of (typically) one
+[Flink JobManager Process](#flink-jobmanager-process) and one or more
```
I have often called this one "master" process, consisting typically of ResourceManager and one or more JobManagers (one per job) and optionally a dispatcher (for session clusters).

For an application cluster: Maser = RM + JM
For as session cluster: Master = RM + Dispatcher + [JM per job]

StephanEwen replied:
I am not sure is "master" process is the best name. It is similar to the old JobManager, but the name is now a bit misleading.

knaufk replied:
We discussed calling it "Flink Master Process", but this would introduce a lot of inconsistencies through out the documentation. So, we decided to call it Flink JobManager Process for now and revisit this/change it, when we rework the Deployment section of the documentation.

StephanEwen replied:
I an skeptical whether terms coined in the glossary ever change easily.

We could also do the following:
  - The per-job component inside the master is actually the "JobMaster" in the code.
  - We could call the process as a while (with RM / Disp. / JobMasters) the "Flink Master Process" and redirect "JobManager" there, mentioning that it was the old term used synonymously.
  - Or we keep the term "JobManager" indefinitely and hope that between JobManager and JobMaster there is not too much confusion.

knaufk replied:
I think, we have three options: 

1. Rename "JobManager Process" to "Flink Master" now in the glossary. This would be in line with the future vision, but will create a lot of inconsistencies (e.g. naming of scripts and commands in the binary distribution), which we can not fix as part of the glossary. Redirect can mitigate this a bit.

2. Stick with "JobManager Process" right now and rename later consistently. 

3. Stick with "JobManager" forever and rename the component to "JobMaster" in accordance to the code. I think, it is too late for this (all FLIP-6 documents talk about "JobManager") and the names would also not be very descriptive and very similar.

I am happy with 1. and 2., but not so much with option 3. 

@tillrohrmann, @StephanEwen: What do you think?

tillrohrmann replied:
I would be in favor of 1. with redirecting deprecated terms to it + creating follow up issues to fix the terminology in the other docs + code.

StephanEwen replied:
My preference would be option (1) as well.

knaufk replied:
Perfect. I'll update the PR accordingly and create a follow-up ticket.

knaufk replied:
Pushed the change and created https://issues.apache.org/jira/browse/FLINK-12639 .

StephanEwen commented:
I have forever been struggling with the term "job" in the context of long running applications. To me a "job" sounds inherently bounded/finite. So a the term "job" would only be used for batch/bounded programs.
Maybe some native English speakers could help here (@sjwiesman or @alpinegizmo)

Flink Programs (processes data streams)
   - can run as continuous applications (on unbounded streams)
   - can run as batch jobs (on bounded streams)

sjwiesman commented:
I agree job sounds finite. I have always taken job in a streaming context to mean a particular version of code / configuration / initial state and application to be those properties over time.

alpinegizmo commented:
I'm not troubled by the term job. To me, a job, whether in Flink or the
work world, is something that may or may not have a pre-defined end point.
A flink batch job is similar to having a contract to build a website; a
streaming job is like being hired as a full-time employee. No job lasts
forever; people and computers die or move on to doing other things.

On Wed, Jul 3, 2019 at 3:27 PM Seth Wiesman <notifications@github.com>
wrote:

> I agree job sounds finite. I have always taken job in a streaming context
> to mean a particular version of code / configuration / initial state and
> application to be those properties over time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/flink/pull/8607?email_source=notifications&email_token=AAAKUWFSQ34U6FSYPUVPDG3P5SSMVA5CNFSM4HSZHHI2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZEN7QA#issuecomment-508092352>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKUWDEKZOTHWNRINQ3ALTP5SSMVANCNFSM4HSZHHIQ>
> .
>

StephanEwen commented:
Thanks for clarifying that. Happy to stick with "job" then.

knaufk commented:
@rmetzger I think all open discussions have been resolved and this PR is ready to be merged.

rmetzger commented:
Thanks for adding a glossary to our documentation. I'm merging this initial version now!

rmetzger commented:
@flinkbot approve all
